### PR TITLE
[SPARK-35382][PYTHON] Fix lambda variable name issues in nested DataFrame functions in Python APIs.

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

This PR fixes the same issue as #32424.

```py
from pyspark.sql.functions import flatten, struct, transform
df = spark.sql("SELECT array(1, 2, 3) as numbers, array('a', 'b', 'c') as letters")
df.select(flatten(
    transform(
        "numbers",
        lambda number: transform(
            "letters",
            lambda letter: struct(number.alias("n"), letter.alias("l"))
        )
    )
).alias("zipped")).show(truncate=False)
```

**Before:**

```
+------------------------------------------------------------------------+
|zipped                                                                  |
+------------------------------------------------------------------------+
|[{a, a}, {b, b}, {c, c}, {a, a}, {b, b}, {c, c}, {a, a}, {b, b}, {c, c}]|
+------------------------------------------------------------------------+
```

**After:**

```
+------------------------------------------------------------------------+
|zipped                                                                  |
+------------------------------------------------------------------------+
|[{1, a}, {1, b}, {1, c}, {2, a}, {2, b}, {2, c}, {3, a}, {3, b}, {3, c}]|
+------------------------------------------------------------------------+
```

### Why are the changes needed?

To produce the correct results.

### Does this PR introduce _any_ user-facing change?

Yes, it fixes the results to be correct as mentioned above.

### How was this patch tested?

Added a unit test as well as manually.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @maropu @zero323 @HyukjinKwon

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -89,7 +89,7 @@ jobs:
       id: sync-branch
       run: |
         apache_spark_ref=`git rev-parse HEAD`
-        git fetch https://github.com/$GITHUB_REPOSITORY.git ${GITHUB_REF##*/}
+        git fetch https://github.com/$GITHUB_REPOSITORY.git ${GITHUB_REF#refs/heads/}
```
Comment (dongjoon-hyun) [inline .github/workflows/build_and_test.yml:92]: Hi, @ueshin . If we need this, please make a separate JIRA issue for this.

Comment (ueshin) [reply .github/workflows/build_and_test.yml:92]: Sure. I tested it here to make sure it works. I'll revert the change and file a separate JIRA later. Thanks!

Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:92]: Thanks for the confirmation! Sure, no problem.

**Comment by maropu** [comment]
The fix itself looks fine. Thank you, @ueshin ~

**Review by HyukjinKwon — APPROVED** [review]
The changes look good (except separating GA fix)

**Review by maropu — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.1.

**Pull request closed without merging**